lib/beaker-puppet/install_utils/puppet5.rb

```diff
@@ -0,0 +1,177 @@
+        packaging_data = build_details[packaging_platform]
+        logger.debug("- #{ packaging_data }, isnil? #{ packaging_data.nil? }")
+        if packaging_data.nil?
+          message = <<-EOF
+            :packaging_platform '#{packaging_platform}' for host '#{host}' not in build details
+            :packaging_platform should be the platform-specific key from this list:
+              #{ build_details.keys }
+          EOF
+          fail_test( message )
+        end
+
+        artifact_buildserver_path   = packaging_data[:artifact]
+        repoconfig_buildserver_path = packaging_data[:repo_config]
+        fail_test('no artifact_buildserver_path found') if artifact_buildserver_path.nil?
+
+        artifact_url    = "#{build_url}/#{artifact_buildserver_path}"
+        repoconfig_url  = "#{build_url}/#{repoconfig_buildserver_path}" unless repoconfig_buildserver_path.nil?
+        artifact_url_correct = link_exists?( artifact_url )
+        logger.debug("- artifact url: '#{artifact_url}'. Exists? #{artifact_url_correct}")
+        fail_test('artifact url built incorrectly') if !artifact_url_correct
+
+        return artifact_url, repoconfig_url
+      end
+
+      # install build artifact on the given host
+      #
+      # @param [Host] host Host to install artifact on
+      # @param [String] project_name Name of project for artifact. Needed for OSX installs
+      # @param [String] artifact_url URL of the project install artifact
+      #
+      # @return nil
+      def install_artifact_on(host, artifact_url, project_name)
+        variant, _, _, _ = host[:platform].to_array
+        onhost_package_file = nil
+        if variant == 'eos'
+          host.get_remote_file( artifact_url )
+          onhost_package_file = File.basename( artifact_url )
+        elsif variant == 'solaris'
+          artifact_filename = File.basename( artifact_url )
+          artifact_folder = File.dirname( artifact_url )
+          fetch_http_file( artifact_folder, artifact_filename, '.' )
+          onhost_package_dir = host.tmpdir( 'puppet_installer' )
+          scp_to host, artifact_filename, onhost_package_dir
+          onhost_package_file = "#{ onhost_package_dir }/#{ artifact_filename }"
+        elsif variant == 'osx'
+          on host, "curl -O #{ artifact_url }"
+          onhost_package_file = "#{ project_name }*"
+        end
+
+        if variant == 'eos'
+          # TODO Will be refactored into {Beaker::Host#install_local_package}
+          #   immediately following this work. The release timing makes it
+          #   necessary to have this here separately for a short while
+          host.install_from_file( onhost_package_file )
+        elsif onhost_package_file
+          if onhost_package_dir
+            host.install_local_package( onhost_package_file, '.' )
+          else
+            host.install_local_package( onhost_package_file )
+          end
+        else
+          host.install_package( artifact_url )
```

Comment:
@melissa yes, this code path will only get called with full artifact URLs when repo_configs aren't set.

In terms of your error question, one of the goals of these changes is that we want beaker to remain dumb & not able to know when repo_configs should & shouldn't be available, so testing for this would be difficult unless we broke that. It seems like the packaging system that generates the repo_configs would be better suited to validating repo_config existence before saying a build is finished.

I can see how validating & behaving nicely on both sides would be a nice experience for the end user, but I think that's how we often end up with poor duplicate copies between beaker & the packaging system. 

Right now the [output of a failure case](https://github.com/puppetlabs/beaker/blob/master/lib/beaker/dsl/helpers/web_helpers.rb#L63-L67) is pretty explicit about what caused it to fail if these are unreachable, so I don't think it's hard to debug. There are some other options for dealing with failure cases, but I think we'll talk about them in the other comment that's on the specific lines in question.